### ISSUE TITLE
loadBedGraphTrackFromURL rewrites the config vars including url

### ISSUE THREAD

**Issue by malcook** [paul-shannon/igvShiny#35]
There appears to be some debugging/testing code left in [Shiny\.addCustomMessageHandler\("fubar",](https://github.com/paul-shannon/igvShiny/blob/96d73672be10356e5dacc9a3df4b6be51fd69032/inst/htmlwidgets/igvShiny.js#L511) which is responsible for rewriting the url to a hardcoded value, as can be seen in the javascript console:

_(text version copied below)_
![image](https://user-images.githubusercontent.com/484282/194477277-2374606d-5e64-44e6-a9c8-1bf6cda69674.png)

Perhaps I should not be working off master?

FWIW: I am trying to suss out how to load a bigbed file by URL and could use some guidance...

```
=== loadBedGraphTrackFromURL
igvShiny.js:7 
{elementID: 'igvShiny_0', trackName: 'meme.ame.all.shuffle', url: 'http://webfs/Sci/SCI-003911-GPFGRIZ/sequenceanalysis/run8/PRLT/meme/ame/shuffle/all.shuffle.bb', color: 'gray', trackHeight: 30, …}
igvShiny.js:521 --- loading bedGraphTrackFromURL
igvShiny.js:522 
{url: 'https://www.encodeproject.org/files/ENCFF000ASF/@@download/ENCFF000ASF.bigWig', name: 'GM12878 H3K4me3', color: 'rgb(200,0,0)', autoscaleGroup: '1'}
autoscaleGroup
: 
"1"
color
: 
"rgb(200,0,0)"
format
: 
"bigwig"
name
: 
"GM12878 H3K4me3"
type
: 
"wig"
url
: 
"https://www.encodeproject.org/files/ENCFF000ASF/@@download/ENCFF000ASF.bigWig"
[[Prototype]]
: 
Object
```

**Comment by paul-shannon** [comment]
Malcolm, 

I will try to borrow a laptop over the weekend here in Weimar, work on the issues you raise. 

If you attach a -small- bigbed file, gzipped, to a new issue on GitHub I will give that a try also.

**Comment by malcook** [comment]
Paul,

I have placed a smallish bigbed at https://research.stowers.org/mec/A.distal.all.bb

Bigbeds are already compressed so I sill not gzip it.

I can load into IGV desktop using "File > Load From URL"

But, there is no hurry at all since I have figured out how to define my own generic `loadTrack` function which formats `...` as JSON payload to config any `igv.js` compatible track.

Like this:

```
loadTrack<-function (session, id, trackName,
                     deleteTracksOfSameName = TRUE, quiet = TRUE,...)
{
    if (!quiet) {
        log("--- igvShiny::loadTrack")
    }
    if (deleteTracksOfSameName) {
        removeTracksByName(session, id, trackName)
    }
    configJSON<-rjson::toJSON(rlang::dots_list(...))
    state[["userAddedTracks"]] <- unique(c(state[["userAddedTracks"]],         trackName))
    msg.to.igv <- list(elementID = id, trackName=trackName,configJSON=configJSON)
    session$sendCustomMessage("loadTrack", msg.to.igv )
}
```
Here is the corresponding javascript message handler:
```
Shiny.addCustomMessageHandler("loadTrack",
   function(message){
       igvshiny_log("=== loadTrack");
       igvshiny_log(message);
       var elementID = message.elementID;
       var configJSON = message.configJSON
       var igvBrowser = document.getElementById(elementID).igvBrowser;
      igvBrowser.loadTrack(configJSON);
      }
);
```
I am in this way able to load my bigBed files with success as:
```
    loadTrack(session, id="igvShiny_0"
              ,trackName='meme.ame.all.shuffle'
              ,url='https://research.stowers.org/mec/A.distal.all.bb'
              ,type='annotation'
              ,displayMode="expanded"
              ,height=300
              ,visibilityWindow=1000000)
```
_but only after disabling CORS security in my browser by running it as_
```
chrome.exe --user-data-dir="C://Chrome dev session" --disable-web-security
```
since that webserve enforces CORS.

So, again, no hurry on this issue.  

However, if you can address #34 I would be much obliged.

Enjoy your travels.

**Comment by gladkia** [comment]
Hi @malcook,

What's the status of this issue? Is it still valid? Thanks for your help.

**Comment by malcook** [comment]
Hi - 

The issue still holds, namely that "There appears to be some debugging/testing code left in [Shiny.addCustomMessageHandler("fubar",](https://github.com/paul-shannon/igvShiny/blob/96d73672be10356e5dacc9a3df4b6be51fd69032/inst/htmlwidgets/igvShiny.js#L511) which is responsible for rewriting the url to a hardcoded value" 

Also, I have been using my [proposed addition ](https://github.com/gladkia/igvShiny/issues/35#issuecomment-1271891090) and still think this project should consider adopting it. What do you think?

**Comment by gladkia** [comment]
Thanks for the detailed report, @malcook — I went back through this against current `master` (v1.9.5).

**The reported bug is fixed.** The hardcoded-URL rewrite came from a leftover debug handler that had been registered under the name `"fubar"` in place of `"loadBedGraphTrackFromURL"`, and which overwrote the incoming config with a hardcoded ENCODE bigWig URL:

```js
config = {url: 'https://www.encodeproject.org/files/ENCFF000ASF/@@download/ENCFF000ASF.bigWig', ...}
```

That was resolved in commit `6858131` ("bedGraph (bigWig) from url now works", 2022-10-28), which restored the handler name to `loadBedGraphTrackFromURL` and removed the hardcoded config override. `6858131` is part of the maintained history here.

On the 2024 follow-up: the permalink in that comment points to `paul-shannon/igvShiny@96d73672` (2022-09-29), which is pinned to a commit from *before* the fix, so it still shows the old `fubar` handler. The maintained tree here had already been corrected by then — as of Jan 2024 the handler was already `loadBedGraphTrackFromURL` with no hardcoded URL.

Re-verified on current `master`:

- No `"fubar"` handler and no active hardcoded track URL remain in `inst/htmlwidgets/igvShiny.js`.
- `loadBedGraphTrackFromURL` builds its config from the message and passes the caller-supplied `url` straight through — nothing rewrites it.
- Config merging is guarded on both sides of the R ↔ JS bridge: JS `mergeExtraParameters()` only copies message keys **not** already present in the config (so `url` / `type` / `name` can't be clobbered), and R `.sanitizeAndMergeOptions()` gives explicit function arguments priority over user-supplied `trackConfig`, warns on conflicts, and filters unknown keys against an allowlist.

So the "rewrites the config vars including url" behaviour no longer reproduces.

On your generic `loadTrack` idea — passing an arbitrary igv.js track config straight through (great for e.g. bigBed by URL) is a genuinely useful addition and still of interest; it also lines up with the API work planned under the current ISC grant. Since that's an enhancement rather than this bug, I've split it into its own issue (#111) so it doesn't get lost. (Re CORS: that's a property of the hosting server, not igvShiny — the track loads fine once the server sends the appropriate `Access-Control-Allow-Origin` headers.)

Closing this as resolved. Thanks again for the clear write-up and the reproducer.
